### PR TITLE
feat: [PL-40597]: make error message clear when jwt token has expired

### PR DESCRIPTION
**Pull request by sahibaCCM**
## Describe your changes
Earlier when the token has expired for an invite we where giving an error message "We couldn’t find an invitation matching the email address you entered. Please search your email for an invitation from Harness or contact your admin." Which was miss leading for user.
Now we have changed it to give the message suggesting expiration of the invite that is "This invitation URL has expired. Please request for a new invitation from your admin."

Tested for:
-> When the token is invalid 
-> when the token has expired
-> when token is valid
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated t

https://github.com/harness/harness-core/assets/112424209/501a1493-9b56-4d92-a5c6-d9aa29b1ae43

his PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)

### CODE REVIEW

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger messagecheck

**Comment by sahibaCCM** [comment]
trigger messagecheck

**Comment by sahibaCCM** [comment]
trigger uts

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 567 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2HWw4HkwKAaj0qli:-NbK2HWw4HkwKAaj0qlj:bnj7hl0) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L567)):*
> ```Java
>         throw new JWTDecodeException(INVALID_JWT_TOKEN);
>       } else if (TOKEN_EXPIRED.equals(e.getMessage())) {
>         throw new InvalidClaimException(TOKEN_EXPIRED);
> ```

same as above

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 564 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2Twd8NLkXcnANvL0:-NbK2Twd8NLkXcnANvL1:b-3lue06) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L564)):*
> ```Java
>       inviteIdOptional = getInviteIdFromToken(jwtToken);
>     } catch (InvalidRequestException e) {
>       if (INVALID_JWT_TOKEN.equals(e.getMessage())) {
> ```

Handling on basis of error message is very risky. If someone changes error messages (which are bound to change with time for improvements and generally not considered as contract), then all these logics will fail.

**Review by sahibaCCM — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 4 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 564 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2Twd8NLkXcnANvL0:-NbKVbVlAe9WF1nH9o6T:b-3q73d2) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L564)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

Handling on basis of error message is very risky. If someone changes error messages (which are bound to change with time for improvements and generally not considered as contract), then all these logics will fail.
</blockquote></details>

We have moved the error to 
 'ErrorMessageConstants', so changes would reflect in all places and hence logic would stay intact.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 565 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK1z3k6wCRGq54lBoX:-NbKW-3-BM6Twl5c8UCI:b-bs6ejo) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L565)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This is a backward incompatible change, isn't it? Earlier we weren't throwing any exception but rather processing it, but now we throw it right away. If there's any more callers other than the one where its handled, then it will break the usual flow.
</blockquote></details>

Checked for all the callers, we would benefit from these changes as now we'll have more clarity on what failed in other callers.

**Comment by sahibaCCM** [comment]
trigger uts

**Review by mohitg0795 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 4 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 564 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2Twd8NLkXcnANvL0:-NbKhikz7vKaz12pGm8J:b7mtz4t) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L564)):*
<details><summary><i>Previously, sahibaCCM (Sahiba Khan) wrote…</i></summary><blockquote>

We have moved the error to 
 'ErrorMessageConstants', so changes would reflect in all places and hence logic would stay intact.
</blockquote></details>

This is bad coding style IMO. I agree to your point but thats just a hack of doing things instead of doing it the right way. If the team agrees on it, I shall approve it from my end, no issues.

**Review by mohitg0795 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 4 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 565 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK1z3k6wCRGq54lBoX:-NbKi-bp3Lu1SzGDYwgB:b3dhnw2) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L565)):*
<details><summary><i>Previously, sahibaCCM (Sahiba Khan) wrote…</i></summary><blockquote>

Checked for all the callers, we would benefit from these changes as now we'll have more clarity on what failed in other callers. 
</blockquote></details>

I am bit skeptical how it isn't affecting even a single caller (assuming there are some N callers). Instead of returning response which used to happen in all cases, now it even throws exceptions in some cases. This is surely a breaking change, but I would leave it upto you to examine it further and decide. 
For now, if its checked that its a backward compatible change, resolving the comment from my end.

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger commonchecks

**Comment by sahibaCCM** [comment]
trigger codebasehash

**Comment by sahibaCCM** [comment]
trigger pmd

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by sahibaCCM** [comment]
trigger uts

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 562 at r6](https://reviewable.io/reviews/harness/harness-core/51467#-NbP7hwbAuK5Oie5lTFa:-NbP7hwbAuK5Oie5lTFb:b2qehgk) ([raw file](https://github.com/harness/harness-core/blob/7fe1c61d2ee0a717285989d32304445cf53d16a2/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L562)):*
> ```Java
>     } catch (InvalidRequestException e) {
>       log.error("Invalid invite JWT token", e);
>       if (TOKEN_EXPIRED.equals(e.getMessage()) && ErrorCode.EXPIRED_TOKEN.equals(e.getCode())) {
> ```

nit : ErrorCode.EXPIRED_TOKEN == e.getCode(), thanks



_Code quote:_
```Java
ErrorCode.EXPIRED_TOKEN.equals(e.getCode())
```

**Comment by pbarapatre10** [comment]
*[`120-ng-manager/src/test/java/io/harness/ng/core/invites/api/InviteServiceImplTest.java` line 441 at r6](https://reviewable.io/reviews/harness/harness-core/51467#-NbP7y_jA1D-j8wpqIhm:-NbP7y_jA1D-j8wpqIhn:boqsl4w) ([raw file](https://github.com/harness/harness-core/blob/7fe1c61d2ee0a717285989d32304445cf53d16a2/120-ng-manager/src/test/java/io/harness/ng/core/invites/api/InviteServiceImplTest.java#L441)):*
> ```Java
>     when(jwtGeneratorUtils.verifyJWTToken(any(), any()))
>         .thenThrow(new InvalidRequestException(INVALID_JWT_TOKEN))
>         .thenReturn(Collections.emptyMap());
> ```

nit: why the need to mock and return different-different results from a single mock? 



_Code quote:_
```Java
.thenReturn(Collections.emptyMap());
```

**Review by pbarapatre10 — APPROVED** [review]
> Reviewed 2 of 4 files at r1, 1 of 2 files at r3, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 5 file(s) to review. 1 change requests. 1 of 2 approvals obtained.

**Comment by sahibaCCM** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java [hunk 1]
```diff
@@ -551,6 +559,9 @@ public Optional<Invite> getInviteFromToken(String jwtToken, boolean allowDeleted
       inviteIdOptional = getInviteIdFromToken(jwtToken);
     } catch (InvalidRequestException e) {
       log.error("Invalid invite JWT token", e);
+      if (TOKEN_EXPIRED.equals(e.getMessage()) && ErrorCode.EXPIRED_TOKEN.equals(e.getCode())) {
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java:562]: Should this be OR condition?

File: 980-commons/src/main/java/io/harness/eraro/ErrorMessageConstants.java [hunk 2]
```diff
@@ -10,4 +10,8 @@
 public class ErrorMessageConstants {
   public static final String FORCE_DELETE_SETTING_NOT_ENABLED =
       "Force Delete is not enabled for this account. Please go to Account Resources > Default Settings > Enable Force Delete of Harness Resources to enable this.";
+
+  public static final String INVALID_JWT_TOKEN = "Invalid JWTToken received, failed to decode the token";
```
Comment (meenakshi-harness) [inline 980-commons/src/main/java/io/harness/eraro/ErrorMessageConstants.java:14]: space between JWT Token

File: 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java (hunk 1 above)
Comment (meenakshi-harness) [reply 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java:562]: Also depending on error message is not right way, we can just check on code I think

File: 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java [hunk 3]
```diff
@@ -551,6 +559,9 @@ public Optional<Invite> getInviteFromToken(String jwtToken, boolean allowDeleted
...
+        throw new ExpiredTokenException(TOKEN_EXPIRED, WingsException.USER);
+      }
```
Comment (meenakshi-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java:564]: What happens if ErrorCode.EXPIRED_TOKEN.equals(e.getCode()) is not true?
We should throw actual exception

File: 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java (hunk 1 above)
Comment (sahibaCCM) [reply 120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java:562]: It is an and condition only, We are making sure that the exception is originating from JWT verify.

File: 980-commons/src/main/java/io/harness/eraro/ErrorMessageConstants.java (hunk 2 above)
Comment (sahibaCCM) [reply 980-commons/src/main/java/io/harness/eraro/ErrorMessageConstants.java:14]: Done

**Comment by sahibaCCM** [comment]
trigger uts

**Review by sahibaCCM — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/51467)* status: 5 file(s) to review. 1 change requests. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 564 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2Twd8NLkXcnANvL0:-NbTXv1kD6hVv16jBLam:bj9v81k) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L564)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

This is bad coding style IMO. I agree to your point but thats just a hack of doing things instead of doing it the right way. If the team agrees on it, I shall approve it from my end, no issues.
</blockquote></details>

Discussed with the team, changes look good.
cc: @pbarapatre10 
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 565 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2AdFBF9qYoNjgjSA:-NbTZ_Wu6998ktvDzq3W:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L565)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

We should throw better exception from here instead of jwt token exceptions. eg: "InviteTokenExpiredException" etc.
</blockquote></details>

done
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 565 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK1z3k6wCRGq54lBoX:-NbTZa_wARPwN62vHa6C:b-k2vv01) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L565)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

I am bit skeptical how it isn't affecting even a single caller (assuming there are some N callers). Instead of returning response which used to happen in all cases, now it even throws exceptions in some cases. This is surely a breaking change, but I would leave it upto you to examine it further and decide. 
For now, if its checked that its a backward compatible change, resolving the comment from my end.
</blockquote></details>

Discussed with @pbarapatre10 and we fine with the change.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java` line 567 at r2](https://reviewable.io/reviews/harness/harness-core/51467#-NbK2HWw4HkwKAaj0qli:-NbT_DVcA5DAW72GGrQF:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/725ae2e53180b1fbadedbd713391fcb02270580a/120-ng-manager/src/main/java/io/harness/ng/core/invites/api/impl/InviteServiceImpl.java#L567)):*
<details><summary><i>Previously, mohitg0795 (Mohit Garg) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done

**Pull request closed without merging**